### PR TITLE
Logging and history

### PR DESCRIPTION
**Pull request by matthiasclasen**
Add some information about transaction results to the systemd journal.

### CODE REVIEW

**Comment by matthiasclasen** [comment]
The last commit adds a history command that presents the log entries. This is still a bit provisional. It needs options to limit the data (both by date and by installation / ref / etc. It should also use the --columns facility that is still not merged, in order to allow flexible display of more fields. The date formatting needs some attention too.

**Comment by matthiasclasen** [comment]
More work is in the column-logging branch

**Comment by matthiasclasen** [comment]
One question here is if this logging at the right level in flatpak-transaction.c, or rather belongs in flatpak-dir.c

**Comment by matthiasclasen** [comment]
I've done these changes on the column-logging branch now.

**Comment by matthiasclasen** [comment]
Example history output:

[log.txt](https://github.com/flatpak/flatpak/files/2409246/log.txt)

You can see how the pull handling for a system install goes through first pulling into a local repo (as the user), then doing a local pull from there (as root) and then deploying (as root).

You can also see that we correctly show how the system helper operates as root on behalf of mclasen, and triggered by the flatpak binary.

**Comment by matthiasclasen** [comment]
Example history output:

[log.txt](https://github.com/flatpak/flatpak/files/2409252/log.txt)


You can see how the pull handling for a system install goes through first pulling into a local repo (as the user), then doing a local pull from there (as root) and then deploying (as root).

You can also see that we correctly show how the system helper operates as root on behalf of mclasen, and triggered by the flatpak binary.

File: common/flatpak-dir.c [hunk 1]
```diff
@@ -8189,7 +8247,7 @@ flatpak_dir_uninstall (FlatpakDir                 *self,
   parts = flatpak_decompose_ref (ref, error);
   if (parts == NULL)
     return FALSE;
-  name = parts[1];
+  app = parts[1];
```
Comment (alexlarsson) [inline common/flatpak-dir.c]: Why did this change to app? It could be a runtime too, no?

File: common/flatpak-utils.c [hunk 2]
```diff
@@ -5792,3 +5799,57 @@ flatpak_utils_ascii_string_to_unsigned (const gchar  *str,
...
+
+  va_start (args, format);
+  g_vsnprintf (message, sizeof (message), format, args);
+  va_end (args);
+
+  if (opid)
+    sd_journal_send ("MESSAGE_ID=" FLATPAK_MESSAGE_ID,
+                     "PRIORITY=5",
+                     "FLATPAK_VERSION=" PACKAGE_VERSION,
+                     "OBJECT_PID=%d", opid,
```
Comment (alexlarsson) [inline common/flatpak-utils.c]: Does "OBJECT_PID" have a pre-defined meaning? Its a bit weird otherwise.

File: system-helper/flatpak-system-helper.c [hunk 3]
```diff
@@ -172,6 +221,8 @@ handle_deploy (FlatpakSystemHelper   *object,
       return TRUE;
     }
 
+  g_object_set_data (G_OBJECT (system), "object-pid", GINT_TO_POINTER (get_sender_pid (invocation)));
```
Comment (alexlarsson) [inline system-helper/flatpak-system-helper.c]: This is just too ugly, i think we should add flatpak_dir_set/get_source_pid().

**Comment by alexlarsson** [comment]
There are some issues making this a bit confusing imho. For example, if multiple users install per-user app they all appear mixed up.

A system install is correct:

```
Sep 23 18:33:17 deploy install org.gnome.Characters        x86_64       stable system                                    flathub                                   962a0c928fd1 root (mclasen) flatpak-system-helper (flatpak) 1.0.2  
Sep 23 18:33:17 pull local     org.gnome.Characters        x86_64       stable system                                    /var/tmp/flatpak-cache-BHWYPZ/repo-35vO8z 962a0c928fd1 root (mclasen) flatpak-system-helper (flatpak) 1.0.2  
Sep 23 18:33:15 pull           org.gnome.Characters        x86_64       stable /var/tmp/flatpak-cache-BHWYPZ/repo-35vO8z flathub                                   962a0c928fd1 mclasen        flatpak                         1.0.2  
```

But maybe not super obvious if you don't know how flatpak works. (Also the reversed order makes it even more confusing.)

File: common/flatpak-utils.c (hunk 2 above)
Comment (matthiasclasen) [reply common/flatpak-utils.c]: It is defined by the journal, for 'acting on behalf of somebody else' - setting it makes journald pick up a bunch of other fields. See systemd.journal-fields(7)

File: common/flatpak-dir.c (hunk 1 above)
Comment (matthiasclasen) [reply common/flatpak-dir.c]: I guess I wanted to keep 'name' in sync with the other changes in this commit. But I can change it back, no problem

File: common/flatpak-utils.c (hunk 2 above)
Comment (alexlarsson) [reply common/flatpak-utils.c]: Ok, that seems useful.

**Comment by matthiasclasen** [comment]
I've rewritten things a bit - there's now FlatpakDir api to set the source pid and the history command logs oldest-first by default (with a --reverse option to flip it).

File: common/flatpak-dir.c [hunk 4]
```diff
@@ -12586,3 +12681,69 @@ flatpak_dir_get_source_pid (FlatpakDir *self)
...
+  len = g_snprintf (message, sizeof (message), "%s: ", installation);
+
+  va_start (args, format);
+  g_vsnprintf (message + len, sizeof (message) - len, format, args);
+  va_end (args);
+
+  /* See systemd.journal-fields(7) for the meaning of the
+   * standard fields we use, in particular OBJECT_PID
+   */
+  if (source_pid)
```
Comment (matthiasclasen) [inline common/flatpak-dir.c]: The duplication here is because I'm not 100% sure if setting OBJECT_PID to 0 is going to confuse journald.

File: common/flatpak-dir.c [hunk 5]
```diff
@@ -8290,8 +8368,12 @@ flatpak_dir_uninstall (FlatpakDir                 *self,
     {
       g_set_error (error, FLATPAK_ERROR, FLATPAK_ERROR_NOT_INSTALLED,
                    _("%s branch %s is not installed"), name, parts[3]);
+      return FALSE;
```
Comment (alexlarsson) [inline common/flatpak-dir.c]: Can you make this a separate commit so we can backport this to 1.0.x?

File: common/flatpak-dir.c [hunk 6]
```diff
@@ -7218,6 +7292,10 @@ flatpak_dir_deploy_update (FlatpakDir   *self,
 
   flatpak_dir_cleanup_removed (self, cancellable, NULL);
 
+  commit = flatpak_dir_read_active (self, ref, cancellable);
+  flatpak_dir_log (self, "deploy update", old_origin, ref, commit, NULL,
+                   "Updated %s from %s", ref, old_origin);
```
Comment (alexlarsson) [inline common/flatpak-dir.c]: I wonder if this should log the old deploy commit too. I realize this is duplicated if you have an infinite log, but that may not always be true, and even then looking up the previous version might be somewhat of a pain if it was a long time ago.

Comment (matthiasclasen) [reply common/flatpak-dir.c]: sure, can do

File: common/flatpak-dir.c (hunk 5 above)
Comment (matthiasclasen) [reply common/flatpak-dir.c]: I will

**Comment by matthiasclasen** [comment]
hmm, not sure whats going with the tests. didn't touch them

**Comment by alexlarsson** [comment]
@rh-atomic-bot r+

**Comment by alexlarsson** [comment]
@rh-atomic-bot retry
